File: integrations/hermes-atlas/atlas/cognitive_client.py
"""Thin managed HTTP client for the authoritative Atlas cognitive service."""

from __future__ import annotations

import hashlib
import json
import os
import secrets
import subprocess
import sys
import tempfile
import time
import urllib.error
import urllib.parse
import urllib.request
from pathlib import Path
from typing import Any

SERVICE_VERSION = "0.1.0"
STATE_TEMP_STALE_SECONDS = 3600


class CognitiveServiceError(RuntimeError):
    def __init__(self, message: str, *, code: str = "service_error", status: int = 0):
        super().__init__(message)
        self.code = code
        self.status = status


class CognitiveServiceUnavailable(CognitiveServiceError):
    pass


def _canonical(value: Any) -> str:
    return json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=True)


class ManagedCognitiveClient:
    """Profile-scoped transport and sidecar lifecycle; contains no cognition."""

    def __init__(
        self,
        *,
        scope_id: str,
        hermes_home: Path,
        data_dir: Path,
        configured_url: str = "",
        expected_scope: str = "",
        timeout_seconds: float = 5.0,
    ) -> None:
        self.scope_id = scope_id
        self.hermes_home = hermes_home
        self.data_dir = data_dir
        self.timeout_seconds = timeout_seconds
        self._process: subprocess.Popen[Any] | None = None
        self._log_handle: Any = None
        self._managed = not configured_url.strip()
        self._owner_instance = secrets.token_hex(16)
        self._owns_running_instance = False

        if self._managed:
            state = self._managed_state()
            self.base_url = f"http://127.0.0.1:{state['port']}"
            self.token = str(state["token"])
            self.expected_scope = scope_id
            self._database = self.data_dir / f"cognitive-service-{scope_id[:16]}.db"
        else:
            self.base_url = self._validate_local_url(configured_url)
            self.token = os.environ.get("ATLAS_COGNITIVE_TOKEN", "")
            if len(self.token) < 32:
                raise CognitiveServiceError(
                    "Configured cognitive_url requires ATLAS_COGNITIVE_TOKEN with at least 32 characters"
                )
            if not expected_scope:
                raise CognitiveServiceError(
                    "Configured cognitive_url requires cognitive_expected_scope"
                )
            if expected_scope != scope_id:
                raise CognitiveServiceError(
                    "cognitive_expected_scope must equal this Hermes profile's internal scope"
                )
            self.expected_scope = expected_scope
            self._database = None

    @staticmethod
    def _validate_local_url(value: str) -> str:
        parsed = urllib.parse.urlparse(value.strip())
        if parsed.scheme != "http" or parsed.hostname not in {"127.0.0.1", "localhost", "::1"}:
            raise CognitiveServiceError("cognitive_url must be an HTTP loopback address")
        if parsed.username or parsed.password or parsed.query or parsed.fragment:
            raise CognitiveServiceError("cognitive_url cannot contain credentials, query, or fragment")
        if parsed.path not in {"", "/"}:
            raise CognitiveServiceError("cognitive_url must not contain a path")
        return value.rstrip("/")

    def _managed_state(self) -> dict[str, Any]:
        state_dir = self.hermes_home / "atlas" / "service"
        state_dir.mkdir(mode=0o700, parents=True, exist_ok=True)
        if os.name != "nt":
            os.chmod(state_dir, 0o700)
        state_path = state_dir / f"{self.scope_id}.json"
        temporary_prefix = f".{self.scope_id}."
        instance_key = f"{self.scope_id}\0{self.hermes_home.resolve()}"
        digest = hashlib.sha256(instance_key.encode("utf-8")).digest()
        port = 20000 + int.from_bytes(digest[:8], "big") % 30000

        def publish(payload: dict[str, Any], *, overwrite: bool) -> None:
            temporary_path: Path | None = None
            try:
                if not overwrite and state_path.exists():
                    return
                descriptor, temporary_name = tempfile.mkstemp(
                    prefix=temporary_prefix, suffix=".tmp", dir=state_dir
                )
                temporary_path = Path(temporary_name)
                if os.name != "nt":
                    os.fchmod(descriptor, 0o600)
                with os.fdopen(descriptor, "w", encoding="utf-8") as handle:
                    handle.write(json.dumps(payload, sort_keys=True) + "\n")
                    handle.flush()
                    os.fsync(handle.fileno())
                if overwrite:
                    os.replace(temporary_path, state_path)
                    temporary_path = None
                else:
                    try:
                        os.link(temporary_path, state_path)
                    except OSError as exc:
                        if not state_path.exists():
                            raise CognitiveServiceError(
                                "Managed state requires same-directory atomic hard-link support"
                            ) from exc
                if os.name != "nt":
                    os.chmod(state_path, 0o600)
                try:
                    directory_descriptor = os.open(state_dir, os.O_RDONLY)
                    try:
                        os.fsync(directory_descriptor)
                    finally:
                        os.close(directory_descriptor)
                except OSError:
                    pass
            finally:
                if temporary_path is not None:
                    temporary_path.unlink(missing_ok=True)

        def clean_temporary_debris() -> None:
            cutoff = time.time() - STATE_TEMP_STALE_SECONDS
            for candidate in state_dir.glob(f"{temporary_prefix}*.tmp"):
                try:
                    if candidate.stat().st_mtime <= cutoff:
                        candidate.unlink(missing_ok=True)
                except OSError:
                    continue

        if not state_path.exists():
            publish(
                {
                    "scope_id": self.scope_id,
                    "port": port,
                    "token": secrets.token_urlsafe(32),
                    "service_version": SERVICE_VERSION,
                },
                overwrite=False,
            )
        try:
            state = json.loads(state_path.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError) as exc:
            raise CognitiveServiceError(f"Invalid managed cognitive state file: {state_path}") from exc
        if state.get("scope_id") != self.scope_id:
            raise CognitiveServiceError(f"Managed cognitive state does not match profile: {state_path}")
        if len(str(state.get("token") or "")) < 32:
            raise CognitiveServiceError(f"Managed cognitive token is invalid: {state_path}")
        if state.get("port") != port:
            state["port"] = port
            state["service_version"] = SERVICE_VERSION
            publish(state, overwrite=True)
        clean_temporary_debris()
        return state

    @staticmethod
    def operation_key(prefix: str, session_id: str, payload: dict[str, Any]) -> str:
        raw = _canonical({"prefix": prefix, "session_id": session_id, "payload": payload})
        return f"hermes-{prefix}-" + hashlib.sha256(raw.encode("utf-8")).hexdigest()

    @staticmethod
    def memory_id(idempotency_key: str) -> str:
        digest = hashlib.sha256(idempotency_key.encode("utf-8")).hexdigest()
        return f"hermes-{digest[:32]}"

    def _service_root(self) -> Path:
        source = Path(__file__).resolve()
        candidates = [source.parent / "cognitive-service", source.parents[2] / "cognitive-service"]
        root = next((item for item in candidates if (item / "server.py").is_file()), None)
        if root is None:
            raise CognitiveServiceUnavailable(
                "Managed cognitive service assets are missing; reinstall the Atlas Hermes package"
            )
        return root

    def _launch(self) -> None:
        if not self._managed or self._database is None:
            raise CognitiveServiceUnavailable(
                f"Atlas cognitive service is unavailable at {self.base_url}"
            )
        if self._process is not None and self._process.poll() is not None:
            self._process.wait()
            self._process = None
        if self._log_handle is not None:
            self._log_handle.close()
            self._log_handle = None
        root = self._service_root()
        self._database.parent.mkdir(parents=True, exist_ok=True)
        log_path = self.hermes_home / "atlas" / "service" / f"{self.scope_id}.log"
        self._log_handle = log_path.open("ab", buffering=0)
        command = [
            sys.executable,
            str(root / "server.py"),
            "--db",
            str(self._database),
            "--scope",
            self.scope_id,
            "--port",
            str(urllib.parse.urlparse(self.base_url).port),
            "--owner-instance",
            self._owner_instance,
            "--parent-pid",
            str(os.getpid()),
        ]
        environment = os.environ.copy()
        environment["ATLAS_COGNITIVE_TOKEN"] = self.token
        options: dict[str, Any] = {
            "cwd": str(root),
            "env": environment,
            "stdin": subprocess.DEVNULL,
            "stdout": self._log_handle,
            "stderr": subprocess.STDOUT,
        }
        if os.name == "nt":
            options["creationflags"] = (
                subprocess.CREATE_NEW_PROCESS_GROUP | subprocess.DETACHED_PROCESS
            )
        else:
            options["start_new_session"] = True
        self._process = subprocess.Popen(command, **options)
        self._owns_running_instance = True

    def _health(self) -> dict[str, Any]:
        health = self._raw("GET", "/v1/health")
        if health.get("scope_id") != self.expected_scope:
            raise CognitiveServiceError(
                "Cognitive service scope mismatch; refusing cross-profile access"
            )
        if health.get("service_version") != SERVICE_VERSION:
            raise CognitiveServiceError(
                f"Cognitive service version mismatch: expected {SERVICE_VERSION}, "
                f"received {health.get('service_version')!r}"
            )
        return health

    def ensure_available(self) -> dict[str, Any]:
        try:
            health = self._health()
        except CognitiveServiceUnavailable:
            if not self._managed:
                raise
        else:
            if not self._managed:
                return health
            owner = health.get("managed_owner") or {}
            if owner.get("instance_id") == self._owner_instance:
                self._owns_running_instance = True
                return health
            if not owner.get("instance_id"):
                raise CognitiveServiceError(
                    "Managed endpoint has no owner metadata; refusing unsafe attachment"
                )
            self._owns_running_instance = False
            return health
        self._launch()
        deadline = time.monotonic() + 5.0
        while time.monotonic() < deadline:
            try:
                health = self._health()
                owner = health.get("managed_owner") or {}
                if owner.get("instance_id") != self._owner_instance:
                    # A concurrent healthy launcher won the stable scoped port.
                    # Attach without ownership; only that launcher's watchdog or
                    # clean shutdown may stop the shared service.
                    self._owns_running_instance = False
                return health
            except CognitiveServiceUnavailable:
                time.sleep(0.05)
        raise CognitiveServiceUnavailable(
            f"Managed cognitive service did not become ready at {self.base_url}; "
            f"inspect {self.hermes_home / 'atlas' / 'service' / (self.scope_id + '.log')}"
        )

    def _stop_managed_instance(self, owner_instance: str) -> None:
        self._raw(
            "POST",
            "/v1/control/shutdown",
            {"owner_instance": owner_instance},
        )
        deadline = time.monotonic() + 5.0
        while time.monotonic() < deadline:
            try:
                self._raw("GET", "/v1/health")
            except CognitiveServiceUnavailable:
                return
            time.sleep(0.05)
        raise CognitiveServiceError("Managed cognitive service did not stop for reclamation")

    def _raw(self, method: str, path: str, body: dict[str, Any] | None = None) -> Any:
        encoded = None if body is None else _canonical(body).encode("utf-8")
        request = urllib.request.Request(
            self.base_url + path,
            data=encoded,
            method=method,
            headers={
                "Authorization": f"Bearer {self.token}",
                "Accept": "application/json",
                **({"Content-Type": "application/json"} if encoded is not None else {}),
            },
        )
        try:
            with urllib.request.urlopen(request, timeout=self.timeout_seconds) as response:
                envelope = json.loads(response.read())
        except urllib.error.HTTPError as exc:
            try:
                envelope = json.loads(exc.read())
                error = envelope.get("error") or {}
            except (json.JSONDecodeError, AttributeError):
                error = {}
            raise CognitiveServiceError(
                str(error.get("message") or f"Cognitive service HTTP {exc.code}"),
                code=str(error.get("code") or "http_error"),
                status=exc.code,
            ) from exc
        except (urllib.error.URLError, TimeoutError, OSError) as exc:
            raise CognitiveServiceUnavailable(
                f"Atlas cognitive service is unavailable at {self.base_url}"
            ) from exc
        if not isinstance(envelope, dict) or envelope.get("api_version") != "v1":
            raise CognitiveServiceError("Cognitive service returned an invalid v1 envelope")
        if envelope.get("ok") is not True:
            error = envelope.get("error") or {}
            raise CognitiveServiceError(
                str(error.get("message") or "Cognitive service request failed"),
                code=str(error.get("code") or "service_error"),
            )
        return envelope.get("data")

    def request(self, method: str, path: str, body: dict[str, Any] | None = None) -> Any:
        self.ensure_available()
        try:
            return self._raw(method, path, body)
        except CognitiveServiceUnavailable:
            if not self._managed:
                raise
            # The healthy owner may exit after the health check but before the
            # request. Fail over and replay this transport-safe request once.
            self._owns_running_instance = False
            self.ensure_available()
            return self._raw(method, path, body)

    def get(self, memory_id: str) -> dict[str, Any] | None:
        path = "/v1/items/get?" + urllib.parse.urlencode({"root_kref": memory_id})
        try:
            return self.request("GET", path)
        except CognitiveServiceError as exc:
            if exc.code == "not_found":
                return None
            raise

    def audit(self, memory_id: str) -> dict[str, Any] | None:
        path = "/v1/items/audit?" + urllib.parse.urlencode({"root_kref": memory_id})
        try:
            return self.request("GET", path)
        except CognitiveServiceError as exc:
            if exc.code == "not_found":
                return None
            raise

    def search(self, query: str, *, limit: int) -> list[dict[str, Any]]:
        return self.request("POST", "/v1/items/search", {"query": query, "limit": limit})

    def list(self, *, limit: int) -> list[dict[str, Any]]:
        path = "/v1/items/list?" + urllib.parse.urlencode({"limit": limit})
        return self.request("GET", path)

    def create(self, payload: dict[str, Any]) -> dict[str, Any]:
        return self.request("POST", "/v1/items/create", payload)

    def revise(self, payload: dict[str, Any]) -> dict[str, Any]:
        return self.request("POST", "/v1/items/revise", payload)

    def depend(self, payload: dict[str, Any]) -> dict[str, Any]:
        return self.request("POST", "/v1/dependencies", payload)

    def forget(self, payload: dict[str, Any]) -> dict[str, Any]:
        return self.request("POST", "/v1/items/forget", payload)

    def shutdown(self) -> None:
        if self._managed and self._owns_running_instance:
            try:
                self._stop_managed_instance(self._owner_instance)
            except CognitiveServiceUnavailable:
                pass
            except CognitiveServiceError as exc:
                if exc.code != "owner_mismatch":
                    raise
        if self._process is not None:
            try:
                self._process.wait(timeout=3.0)
            except subprocess.TimeoutExpired:
                self._process.terminate()
                try:
                    self._process.wait(timeout=2.0)
                except subprocess.TimeoutExpired:
                    self._process.kill()
                    self._process.wait(timeout=2.0)
        self._process = None
        self._owns_running_instance = False
        if self._log_handle is not None:
            self._log_handle.close()
            self._log_handle = None

File: integrations/openclaw-atlas/src/cognitive-client.ts
import { spawn, type ChildProcess } from "node:child_process";
import {
  closeSync,
  existsSync,
  mkdirSync,
  openSync,
  readFileSync,
  writeFileSync,
} from "node:fs";
import { createHash, randomBytes } from "node:crypto";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const SERVICE_VERSION = "0.1.0";

type ManagedState = {
  port: number;
  scopeId: string;
  serviceVersion: string;
  token: string;
};

type ServiceEnvelope = {
  api_version?: unknown;
  data?: unknown;
  error?: { code?: unknown; message?: unknown };
  ok?: unknown;
};

export class CognitiveServiceError extends Error {
  constructor(
    message: string,
    readonly code = "service_error",
    readonly status = 0,
  ) {
    super(message);
  }
}

export class CognitiveServiceUnavailable extends CognitiveServiceError {}

function sha256(value: string): string {
  return createHash("sha256").update(value).digest("hex");
}

function canonical(value: unknown): string {
  if (Array.isArray(value)) return `[${value.map(canonical).join(",")}]`;
  if (value && typeof value === "object") {
    return `{${Object.keys(value as Record<string, unknown>)
      .sort()
      .map((key) => `${JSON.stringify(key)}:${canonical((value as Record<string, unknown>)[key])}`)
      .join(",")}}`;
  }
  return JSON.stringify(value) ?? "null";
}

function serviceRoot(explicitRoot?: string): string {
  if (explicitRoot) {
    if (!existsSync(join(explicitRoot, "server.py"))) {
      throw new CognitiveServiceUnavailable(
        `Atlas cognitive-service assets are missing from ${explicitRoot}`,
      );
    }
    return explicitRoot;
  }
  const here = dirname(fileURLToPath(import.meta.url));
  const candidates = [
    join(here, "..", "..", "service"),
    join(here, "..", "..", "..", "cognitive-service"),
  ];
  const root = candidates.find((candidate) => existsSync(join(candidate, "server.py")));
  if (!root) {
    throw new CognitiveServiceUnavailable(
      "Atlas cognitive-service assets are missing; reinstall the OpenClaw package",
    );
  }
  return root;
}

function parseState(path: string, expectedScope: string): ManagedState {
  let value: unknown;
  try {
    value = JSON.parse(readFileSync(path, "utf8"));
  } catch {
    throw new CognitiveServiceError(`Invalid managed cognitive state file: ${path}`);
  }
  if (!value || typeof value !== "object") {
    throw new CognitiveServiceError(`Invalid managed cognitive state file: ${path}`);
  }
  const state = value as Partial<ManagedState>;
  if (
    state.scopeId !== expectedScope ||
    !Number.isInteger(state.port) ||
    typeof state.token !== "string" ||
    state.token.length < 32
  ) {
    throw new CognitiveServiceError(`Managed cognitive state does not match scope: ${path}`);
  }
  return state as ManagedState;
}

export class ManagedCognitiveClient {
  readonly baseUrl: string;
  readonly token: string;
  readonly scopeId: string;
  private child: ChildProcess | null = null;
  private launchError: Error | null = null;
  private ownsRunningInstance = false;
  private readonly ownerInstance = randomBytes(16).toString("hex");
  private readonly databasePath: string;
  private readonly logPath: string;

  constructor(
    private readonly dataDir: string,
    scopeKey: string,
    private readonly pythonCommand = process.env.ATLAS_PYTHON ||
      (process.platform === "win32" ? "python" : "python3"),
    private readonly serviceAssetRoot?: string,
    serviceNamespace = "openclaw",
  ) {
    if (!/^[a-z][a-z0-9-]{0,31}$/.test(serviceNamespace)) {
      throw new CognitiveServiceError("Invalid Atlas service namespace");
    }
    const identity = `${dataDir}\0${scopeKey}`;
    this.scopeId = `${serviceNamespace}-${sha256(identity).slice(0, 32)}`;
    const port = 20_000 + (Number.parseInt(sha256(identity).slice(0, 8), 16) % 30_000);
    const serviceDir = join(dataDir, "cognitive", this.scopeId);
    mkdirSync(serviceDir, { recursive: true, mode: 0o700 });
    const statePath = join(serviceDir, "state.json");
    if (!existsSync(statePath)) {
      const state: ManagedState = {
        port,
        scopeId: this.scopeId,
        serviceVersion: SERVICE_VERSION,
        token: randomBytes(32).toString("base64url"),
      };
      try {
        writeFileSync(statePath, `${JSON.stringify(state)}\n`, {
          encoding: "utf8",
          flag: "wx",
          mode: 0o600,
        });
      } catch (error) {
        if (!existsSync(statePath)) throw error;
      }
    }
    const state = parseState(statePath, this.scopeId);
    if (state.port !== port || state.serviceVersion !== SERVICE_VERSION) {
      throw new CognitiveServiceError(
        "Managed cognitive state uses an incompatible port or service version",
      );
    }
    this.baseUrl = `http://127.0.0.1:${state.port}`;
    this.token = state.token;
    this.databasePath = join(serviceDir, "cognitive.sqlite3");
    this.logPath = join(serviceDir, "service.log");
  }

  static operationKey(prefix: string, context: string, payload: unknown): string {
    return `openclaw-${prefix}-${sha256(canonical({ context, payload }))}`;
  }

  static memoryId(idempotencyKey: string): string {
    return `openclaw-${sha256(idempotencyKey).slice(0, 32)}`;
  }

  private async raw(method: string, path: string, body?: unknown): Promise<unknown> {
    let response: Response;
    try {
      response = await fetch(`${this.baseUrl}${path}`, {
        method,
        headers: {
          Accept: "application/json",
          Authorization: `Bearer ${this.token}`,
          ...(body === undefined ? {} : { "Content-Type": "application/json" }),
        },
        ...(body === undefined ? {} : { body: JSON.stringify(body) }),
        signal: AbortSignal.timeout(5_000),
      });
    } catch {
      throw new CognitiveServiceUnavailable(
        `Atlas cognitive service is unavailable at ${this.baseUrl}`,
        "unavailable",
        0,
      );
    }
    let envelope: ServiceEnvelope;
    try {
      envelope = (await response.json()) as ServiceEnvelope;
    } catch {
      throw new CognitiveServiceError("Cognitive service returned invalid JSON", "invalid_response", response.status);
    }
    if (envelope.api_version !== "v1") {
      throw new CognitiveServiceError("Cognitive service returned an invalid v1 envelope");
    }
    if (!response.ok || envelope.ok !== true) {
      throw new CognitiveServiceError(
        typeof envelope.error?.message === "string"
          ? envelope.error.message
          : `Cognitive service HTTP ${response.status}`,
        typeof envelope.error?.code === "string" ? envelope.error.code : "http_error",
        response.status,
      );
    }
    return envelope.data;
  }

  private async health(): Promise<Record<string, unknown>> {
    const health = (await this.raw("GET", "/v1/health")) as Record<string, unknown>;
    if (health.scope_id !== this.scopeId) {
      throw new CognitiveServiceError("Cognitive service scope mismatch; refusing cross-scope access");
    }
    if (health.service_version !== SERVICE_VERSION) {
      throw new CognitiveServiceError(
        `Cognitive service version mismatch: expected ${SERVICE_VERSION}`,
      );
    }
    return health;
  }

  private launch(): void {
    const root = serviceRoot(this.serviceAssetRoot);
    const logFd = openSync(this.logPath, "a", 0o600);
    this.launchError = null;
    try {
      this.child = spawn(
        this.pythonCommand,
        [
          join(root, "server.py"),
          "--db",
          this.databasePath,
          "--scope",
          this.scopeId,
          "--port",
          new URL(this.baseUrl).port,
          "--owner-instance",
          this.ownerInstance,
          "--parent-pid",
          String(process.pid),
        ],
        {
          cwd: root,
          detached: process.platform !== "win32",
          env: { ...process.env, ATLAS_COGNITIVE_TOKEN: this.token },
          stdio: ["ignore", logFd, logFd],
          windowsHide: true,
        },
      );
      // Spawn reports an absent executable asynchronously; retain it so the
      // readiness loop returns the real cause instead of only a timeout.
      this.child.once("error", (error) => {
        this.launchError = error;
      });
    } finally {
      closeSync(logFd);
    }
    this.ownsRunningInstance = true;
  }

  async ensureAvailable(): Promise<Record<string, unknown>> {
    try {
      const health = await this.health();
      const owner = (health.managed_owner ?? {}) as Record<string, unknown>;
      this.ownsRunningInstance = owner.instance_id === this.ownerInstance;
      if (!owner.instance_id) {
        throw new CognitiveServiceError(
          "Managed endpoint has no owner metadata; refusing unsafe attachment",
        );
      }
      return health;
    } catch (error) {
      if (!(error instanceof CognitiveServiceUnavailable)) throw error;
    }
    this.launch();
    const deadline = Date.now() + 5_000;
    while (Date.now() < deadline) {
      if (this.launchError) {
        throw new CognitiveServiceUnavailable(
          `Could not launch Atlas with ${this.pythonCommand}: ${this.launchError.message}`,
        );
      }
      await new Promise((resolve) => setTimeout(resolve, 50));
      try {
        const health = await this.health();
        const owner = (health.managed_owner ?? {}) as Record<string, unknown>;
        this.ownsRunningInstance = owner.instance_id === this.ownerInstance;
        return health;
      } catch (error) {
        if (!(error instanceof CognitiveServiceUnavailable)) throw error;
      }
    }
    throw new CognitiveServiceUnavailable(
      `Managed cognitive service did not become ready; inspect ${this.logPath}`,
    );
  }

  async request(method: string, path: string, body?: unknown): Promise<unknown> {
    await this.ensureAvailable();
    try {
      return await this.raw(method, path, body);
    } catch (error) {
      if (!(error instanceof CognitiveServiceUnavailable)) throw error;
      this.ownsRunningInstance = false;
      await this.ensureAvailable();
      return await this.raw(method, path, body);
    }
  }

  async shutdown(): Promise<void> {
    if (this.ownsRunningInstance) {
      try {
        await this.raw("POST", "/v1/control/shutdown", {
          owner_instance: this.ownerInstance,
        });
      } catch (error) {
        if (!(error instanceof CognitiveServiceUnavailable)) throw error;
      }
    }
    if (this.child && this.child.exitCode === null) {
      await Promise.race([
        new Promise<void>((resolve) => this.child?.once("exit", () => resolve())),
        new Promise<void>((resolve) => setTimeout(resolve, 3_000)),
      ]);
      if (this.child.exitCode === null) this.child.kill("SIGTERM");
    }
    this.child = null;
    this.ownsRunningInstance = false;
  }
}

export class CognitiveClientPool {
  private readonly clients = new Map<string, ManagedCognitiveClient>();

  constructor(
    private readonly dataDir: string,
    private readonly pythonCommand?: string,
    private readonly serviceAssetRoot?: string,
  ) {}

  forScope(scopeKey: string): ManagedCognitiveClient {
    let client = this.clients.get(scopeKey);
    if (!client) {
      client = new ManagedCognitiveClient(
        this.dataDir,
        scopeKey,
        this.pythonCommand,
        this.serviceAssetRoot,
      );
      this.clients.set(scopeKey, client);
    }
    return client;
  }

  async shutdown(): Promise<void> {
    const results = await Promise.allSettled(
      [...this.clients.values()].map((client) => client.shutdown()),
    );
    this.clients.clear();
    const failure = results.find((result) => result.status === "rejected");
    if (failure?.status === "rejected") throw failure.reason;
  }
}
